File: packages/contracts/src/tools/read-tools.ts
import { z } from 'zod';

import { Instant, LocalDate, PhoneE164, PublicSlotId, TimePreference } from './_shared.js';

/**
 * Tools 1–4: the read path. No side effects, no idempotency key, safe to retry
 * (doc 08 §4 — these are the only tools carrying a `backoffPlan`).
 */

// ─── 1. getBusinessInfo ────────────────────────────────────────────────────────

export const GetBusinessInfoInput = z
  .object({
    topic: z
      .enum(['hours', 'location', 'parking', 'contact', 'services_overview', 'policies', 'memberships'])
      .describe(
        'What the caller asked about. Pick the closest match. If nothing fits, do NOT call this — escalate instead.',
      ),
  })
  .strict();

export const GetBusinessInfoOutput = z.object({
  /** Already phrased for speech. The model paraphrases lightly; it must not add facts. */
  answer: z.string(),
  /** False when no approved entry exists — the model must escalate, never improvise. */
  approved: z.boolean(),
});

// ─── 2. lookupCustomer ─────────────────────────────────────────────────────────

export const LookupCustomerInput = z
  .object({
    phone: PhoneE164.describe(
      "The caller's own number from caller ID. You may NOT look up any other number — asking a caller for someone else's number and querying it is not permitted.",
    ),
  })
  .strict();

export const LookupCustomerOutput = z.object({
  found: z.boolean(),
  firstName: z.string().nullable(),
  isMember: z.boolean(),
  /** Present only when found; used to greet by name. Never a full profile. */
  visitCount: z.number().int().nonnegative().nullable(),
  /** True if any booking is upcoming — lets Grace offer reschedule/cancel without a second lookup. */
  hasUpcomingBooking: z.boolean(),
});

// ─── 3. getServicesAndPricing ──────────────────────────────────────────────────

export const GetServicesAndPricingInput = z
  .object({
    query: z
      .string()
      .min(1)
      .max(120)
      .describe(
        "What the caller asked for, in their words, e.g. '60 minute massage' or 'deep tissue'. Pass it through verbatim.",
      ),
    isMember: z
      .boolean()
      .default(false)
      .describe('Set true only if lookupCustomer already confirmed membership. Never assume.'),
  })
  .strict();

export const GetServicesAndPricingOutput = z.object({
  services: z
    .array(
      z.object({
        serviceCode: z.string(),
        name: z.string(),
        durationMinutes: z.number().int().positive(),
        priceCents: z.number().int().nonnegative(),
        memberPriceCents: z.number().int().nonnegative().nullable(),
        depositRequired: z.boolean(),
      }),
    )
    .max(3),
  /** False when the catalogue is unapproved (GATE-04) — Grace must not quote a price. */
  approved: z.boolean(),
});

// ─── 4. checkAvailability ──────────────────────────────────────────────────────

export const CheckAvailabilityInput = z
  .object({
    serviceCode: z
      .string()
      .min(1)
      .describe('Service code, e.g. massage_60. Call getServicesAndPricing first to get it.'),
    preferredDate: LocalDate,
    timePreference: TimePreference.default('any'),
    // `.optional()`, never `.nullable()` — see the note at the bottom of this file.
    providerPreference: z
      .string()
      .optional()
      .describe('Provider name if the caller asked for someone specific. Omit if they did not.'),
    partySize: z.number().int().min(1).max(4).default(1).describe('Number of people. Almost always 1.'),
  })
  .strict();

export const CheckAvailabilityOutput = z.object({
  slots: z
    .array(
      z.object({
        slotId: PublicSlotId,
        startsAt: Instant,
        providerId: z.string(),
        providerName: z.string(),
        priceCents: z.number().int().nonnegative(),
        holdExpiresAt: Instant,
      }),
    )
    /** Hard cap of 3: the prompt forbids reading more than three options aloud. */
    .max(3),
  alternativesAvailable: z.boolean(),
});

/**
 * ⚠️ Tool INPUT schemas must never use `.nullable()`.
 *
 * `zod-to-json-schema` renders a nullable field with any constraint (`.max()`, `.min()`, …)
 * as `anyOf: [{type:"string",…},{type:"null"}]`, and **Vapi rejects `anyOf` in tool
 * parameters** with `400 function.parameters.properties.X.type must be one of …`.
 * Verified against the live API on 2026-08-03.
 *
 * Use `.optional()` instead. It is also the better LLM affordance: the model omits the
 * field rather than reasoning about whether to send an explicit null.
 * `generate-tools.ts` enforces this statically so it can never reach a deploy again.
 *
 * OUTPUT schemas are unaffected — they never leave our process.
 */
export type GetBusinessInfoInput = z.infer<typeof GetBusinessInfoInput>;
export type GetBusinessInfoOutput = z.infer<typeof GetBusinessInfoOutput>;
export type LookupCustomerInput = z.infer<typeof LookupCustomerInput>;
export type LookupCustomerOutput = z.infer<typeof LookupCustomerOutput>;
export type GetServicesAndPricingInput = z.infer<typeof GetServicesAndPricingInput>;
export type GetServicesAndPricingOutput = z.infer<typeof GetServicesAndPricingOutput>;
export type CheckAvailabilityInput = z.infer<typeof CheckAvailabilityInput>;
export type CheckAvailabilityOutput = z.infer<typeof CheckAvailabilityOutput>;

File: src/grace_contracts/tools/read_tools.py
"""Tools 1–4: the read path.

No side effects, no idempotency key, safe to retry — these are the only tools that
carry a ``backoffPlan`` (doc 08 §4).
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import Field

from .shared import LocalDate, PhoneE164, PublicSlotId, TimePreference, ToolInput, ToolOutput

# ── 1. getBusinessInfo ────────────────────────────────────────────────────────


class BusinessTopic(StrEnum):
    HOURS = "hours"
    LOCATION = "location"
    PARKING = "parking"
    CONTACT = "contact"
    SERVICES_OVERVIEW = "services_overview"
    POLICIES = "policies"
    MEMBERSHIPS = "memberships"


class GetBusinessInfoInput(ToolInput):
    topic: BusinessTopic = Field(
        description=(
            "What the caller asked about. Pick the closest match. If nothing fits, "
            "do NOT call this — escalate instead."
        )
    )


class GetBusinessInfoOutput(ToolOutput):
    answer: str = Field(description="Already phrased for speech.")
    approved: bool = Field(
        description="False when no approved entry exists — the model must escalate, never improvise."
    )


# ── 2. lookupCustomer ─────────────────────────────────────────────────────────


class LookupCustomerInput(ToolInput):
    phone: PhoneE164 = Field(
        description=(
            "The caller's own number from caller ID. You may NOT look up any other number — "
            "asking a caller for someone else's number and querying it is not permitted."
        )
    )


class LookupCustomerOutput(ToolOutput):
    found: bool
    first_name: str = ""
    is_member: bool = False
    visit_count: int = 0
    has_upcoming_booking: bool = False


# ── 3. getServicesAndPricing ──────────────────────────────────────────────────


class GetServicesAndPricingInput(ToolInput):
    query: str = Field(
        min_length=1,
        max_length=120,
        description=(
            "What the caller asked for, in their words, e.g. '60 minute massage' or "
            "'deep tissue'. Pass it through verbatim."
        ),
    )
    is_member: bool = Field(
        default=False,
        description="Set true only if lookupCustomer already confirmed membership. Never assume.",
    )


class ServiceOption(ToolOutput):
    service_code: str
    name: str
    duration_minutes: int
    price_cents: int
    member_price_cents: int
    deposit_required: bool


class GetServicesAndPricingOutput(ToolOutput):
    services: list[ServiceOption] = Field(max_length=3)
    approved: bool = Field(
        description="False when the catalogue is unapproved (GATE-04) — Grace must not quote a price."
    )


# ── 4. checkAvailability ──────────────────────────────────────────────────────


class CheckAvailabilityInput(ToolInput):
    service_code: str = Field(
        min_length=1,
        description="Service code, e.g. massage_60. Call getServicesAndPricing first to get it.",
    )
    preferred_date: LocalDate
    time_preference: TimePreference = Field(
        default=TimePreference.ANY,
        description="Rough time of day the caller asked for. Use 'any' if they did not say.",
    )
    # Empty string, not None: an optional-with-constraints field renders as `anyOf`,
    # which Vapi rejects. See the note in shared.py.
    provider_preference: str = Field(
        default="",
        description="Provider name if the caller asked for someone specific. Leave empty if not.",
    )
    party_size: int = Field(default=1, ge=1, le=4, description="Number of people. Almost always 1.")


class AvailableSlot(ToolOutput):
    slot_id: PublicSlotId
    starts_at: str
    provider_id: str
    provider_name: str
    price_cents: int
    hold_expires_at: str


class CheckAvailabilityOutput(ToolOutput):
    # Hard cap of 3: the prompt forbids reading more than three options aloud.
    slots: list[AvailableSlot] = Field(max_length=3)
    alternatives_available: bool
